File: typescript/src/modules/webhook.ts
import { TriggerAgentPayload, TriggerAgentResponse } from '../types';
import { PermissionDeniedError } from './api';

export class WebhookModule {
    private realtimexUrl: string;
    private appName?: string;
    private appId?: string;
    private apiKey?: string;

    constructor(realtimexUrl: string, appName?: string, appId?: string, apiKey?: string) {
        this.realtimexUrl = realtimexUrl.replace(/\/$/, '');
        this.appName = appName;
        this.appId = appId;
        this.apiKey = apiKey;
    }

    /**
     * Request a single permission from Electron via internal API
     */
    private async requestPermission(permission: string): Promise<boolean> {
        try {
            const response = await fetch(`${this.realtimexUrl}/api/local-apps/request-permission`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    app_id: this.appId,
                    app_name: this.appName,
                    permission,
                }),
            });
            const data = await response.json();
            return data.granted === true;
        } catch (error) {
            console.error('[SDK] Permission request failed:', error);
            return false;
        }
    }

    private async request<T>(
        path: string,
        options: RequestInit = {}
    ): Promise<T> {
        const url = `${this.realtimexUrl}${path}`;
        const headers: any = {
            'Content-Type': 'application/json',
            ...(options.headers as any),
        };

        if (this.apiKey) {
            headers['Authorization'] = `Bearer ${this.apiKey}`;
        }
        if (this.appId) {
            headers['x-app-id'] = this.appId;
        }

        const response = await fetch(url, {
            ...options,
            headers,
        });

        const data = await response.json();

        // Handle permission errors
        if (response.status === 403) {
            const errorCode = data.error;
            const permission = data.permission;
            const message = data.message;

            if (errorCode === 'PERMISSION_REQUIRED' && permission) {
                // Try to get permission from user
                const granted = await this.requestPermission(permission);

                if (granted) {
                    // Retry the original request
                    return this.request<T>(path, options);
                } else {
                    throw new PermissionDeniedError(permission, message);
                }
            }

            if (errorCode === 'PERMISSION_DENIED') {
                throw new PermissionDeniedError(permission, message);
            }
        }

        if (!response.ok) {
            throw new Error(data.error || `Request failed: ${response.status}`);
        }
        return data;
    }

    async triggerAgent(payload: TriggerAgentPayload): Promise<TriggerAgentResponse> {
        if (payload.auto_run && (!payload.agent_name || !payload.workspace_slug)) {
            throw new Error('auto_run requires agent_name and workspace_slug');
        }

        return this.request<TriggerAgentResponse>('/webhooks/realtimex', {
            method: 'POST',
            body: JSON.stringify({
                app_name: this.appName,
                app_id: this.appId,
                event: 'trigger-agent',
                payload: {
                    raw_data: payload.raw_data,
                    auto_run: payload.auto_run ?? false,
                    agent_name: payload.agent_name,
                    workspace_slug: payload.workspace_slug,
                    thread_slug: payload.thread_slug,
                    prompt: payload.prompt ?? '',
                },
            }),
        });
    }

    async ping(): Promise<{ success: boolean; app_name: string; message: string }> {
        return this.request<{ success: boolean; app_name: string; message: string }>('/webhooks/realtimex', {
            method: 'POST',
            body: JSON.stringify({
                app_name: this.appName,
                app_id: this.appId,
                event: 'ping'
            }),
        });
    }
}

File: python/realtimex_sdk/webhook.py
import httpx
import os
from typing import Any, Dict, Optional
from .api import PermissionDeniedError


class WebhookModule:
    """Call RealtimeX webhook endpoints with permission handling."""

    def __init__(
        self,
        realtimex_url: str,
        app_name: Optional[str] = None,
        app_id: Optional[str] = None,
        api_key: Optional[str] = None,
    ):
        self.realtimex_url = realtimex_url.rstrip("/")
        self.app_name = app_name or os.environ.get("RTX_APP_NAME", "Local App")
        self.app_id = app_id
        self.api_key = api_key

    async def _request_permission(self, permission: str) -> bool:
        """Request a single permission from Electron via internal API."""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    f"{self.realtimex_url}/api/local-apps/request-permission",
                    json={
                        "app_id": self.app_id,
                        "app_name": self.app_name,
                        "permission": permission,
                    },
                    timeout=60.0  # Long timeout for user interaction
                )
                data = response.json()
                return data.get("granted", False)
        except Exception as e:
            print(f"[SDK] Permission request failed: {e}")
            return False

    async def _handle_response(self, response: httpx.Response, retry_fn) -> Dict[str, Any]:
        """Handle response with permission error handling."""
        if response.status_code == 403:
            data = response.json()
            error_code = data.get("error")
            permission = data.get("permission")
            message = data.get("message")

            if error_code == "PERMISSION_REQUIRED" and permission:
                # Try to get permission from user
                granted = await self._request_permission(permission)

                if granted:
                    # Retry the original request
                    return await retry_fn()
                else:
                    raise PermissionDeniedError(permission, message)

            if error_code == "PERMISSION_DENIED":
                raise PermissionDeniedError(permission, message)

        response.raise_for_status()
        return response.json()

    async def trigger_agent(
        self,
        raw_data: Dict[str, Any],
        auto_run: bool = False,
        agent_name: Optional[str] = None,
        workspace_slug: Optional[str] = None,
        thread_slug: Optional[str] = None,
        prompt: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Trigger agent via webhook with automatic permission handling."""
        if auto_run:
            if not agent_name or not workspace_slug or not thread_slug:
                raise ValueError(
                    "auto_run requires agent_name, workspace_slug, and thread_slug"
                )

        async def do_request():
            headers = {"Content-Type": "application/json"}
            if self.api_key:
                headers["Authorization"] = f"Bearer {self.api_key}"
            if self.app_id:
                headers["x-app-id"] = self.app_id
                
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    f"{self.realtimex_url}/webhooks/realtimex",
                    headers=headers,
                    json={
                        "app_name": self.app_name,
                        "app_id": self.app_id,
                        "event": "trigger-agent",
                        "payload": {
                            "raw_data": raw_data,
                            "auto_run": auto_run,
                            "agent_name": agent_name,
                            "workspace_slug": workspace_slug,
                            "thread_slug": thread_slug,
                            "prompt": prompt,
                        },
                    },
                )
                return await self._handle_response(response, do_request)

        return await do_request()

    async def ping(self) -> Dict[str, Any]:
        """Ping webhook to check connection."""
        async def do_request():
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    f"{self.realtimex_url}/webhooks/realtimex",
                    json={
                        "app_name": self.app_name,
                        "app_id": self.app_id,
                        "event": "ping",
                    },
                )
                return await self._handle_response(response, do_request)

        return await do_request()
